src/org/mozilla/javascript/NativeJavaMap.java

```diff
@@ -0,0 +1,79 @@
+/* -*- Mode: java; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.javascript;
+
+import java.util.Map;
+
+public class NativeJavaMap extends NativeJavaObject {
+
+    private Map<Object, Object> map;
+
+    public NativeJavaMap(Scriptable scope, Object map) {
+        super(scope, map, map.getClass());
+        Class<?> cl = map.getClass();
+        if (!Map.class.isAssignableFrom(cl)) {
+            throw new RuntimeException("java.util.Map expected");
+        }
+        this.map = (Map<Object, Object>) map;
+    }
+
+    @Override
+    public String getClassName() {
+        return "JavaMap";
+    }
+
+
+    @Override
+    public boolean has(String name, Scriptable start) {
+        if (map.containsKey(name)) {
+            return true;
+        }
+        return super.has(name, start);
+    }
+
+    @Override
+    public boolean has(int index, Scriptable start) {
+        if (map.containsKey(index)) {
+            return true;
+        }
+        return super.has(index, start);
+    }
+
+    @Override
+    public Object get(String name, Scriptable start) {
+        if (map.containsKey(name)) {
+            Context cx = Context.getContext();
+            Object obj = map.get(name);
+            return cx.getWrapFactory().wrap(cx, this, obj, obj.getClass());
+        }
+        return super.get(name, start);
+    }
+
+    @Override
+    public Object get(int index, Scriptable start) {
+        if (map.containsKey(index)) {
+            Context cx = Context.getContext();
+            Object obj = map.get(index);
+            return cx.getWrapFactory().wrap(cx, this, obj, obj.getClass());
+        }
+        return super.get(index, start);
+    }
+
+    @Override
+    public void put(String name, Scriptable start, Object value) {
+        map.put(name, Context.jsToJava(value, Object.class));
+    }
+
+    @Override
+    public void put(int index, Scriptable start, Object value) {
+        map.put(index, Context.jsToJava(value, Object.class));
+    }
+
+    @Override
+    public Object[] getIds() {
+        return map.keySet().toArray();
```

Comment:
>This may be over-complication but JavaScript has taught me to that anything is possible...

Haha, i feel your pain ;).

>Is it possible that a user would use this with a map that has keys that are not strings or integers? If so, then it might aid interop with JavaScript if we iterated through the list and ensured that each key is either a string or an integer by calling ScriptRuntime.toInt32 or ScriptRuntime.toString depending on the type, and letting those methods throw a JavaScript exception if something doesn't work (for instance, if they try to use a Symbol or something).

Ok for ensuring the type :).

Reply:
Actually, what about filtering out non String/Integer keys instead of converting them?